pkg/controllers/provisioning/v1alpha1/reallocator/filter.go

```diff
@@ -0,0 +1,50 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package reallocator
+
+import (
+	"context"
+	"fmt"
+	v1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type Filter struct {
+	kubeClient client.Client
+}
+
+// Gets Nodes fitting some underutilization
+// TODO: add predicate
+func (f *Filter) GetUnderUtilizedNodes(ctx context.Context) ([]*v1.Node, error) {
+	// 1. Get all nodes
+	nodeList := &v1.NodeList{}
+
+	// 2. Get nodes and the pods on each node
+	for _, node := range nodeList.Items {
+		pods, err := f.GetPodsOnNode(ctx, node.Name)
+		if err != nil {
+			return []*v1.Node{}, fmt.Errorf("filtering pods on nodes, %w", err)
+		}
+		if len(pods) == 0 {
+			return []*v1.Node{}, nil
+		}
+	}
+	return []*v1.Node{}, nil
+}
+
+// Get Pods scheduled to a node
+func (f *Filter) GetPodsOnNode(ctx context.Context, name string) ([]*v1.Pod, error) {
+	return []*v1.Pod{}, nil
```

Comment:
https://github.com/awslabs/karpenter/blob/main/pkg/controllers/provisioning/v1alpha1/allocator/filter.go#L41

Reply:
Nice catch, I'll add this into the util functions and change that a little.

Reply:
Got it. I'll add it in, thanks for the reference.